lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -593,74 +589,98 @@ SerializedModuleLoaderBase::findModule(ImportPath::Element moduleID,
     if (firstAbsoluteBaseName
         && maybeDiagnoseTargetMismatch(moduleID.Loc, moduleName,
                                        *firstAbsoluteBaseName)) {
-      return false;
+      return SearchResult::Error;
     } else {
-      return None;
+      return SearchResult::NotFound;
     }
   };
 
-  auto result = forEachModuleSearchPath(
-      Ctx,
-      [&](StringRef path, SearchPathKind Kind,
-          bool isSystem) -> Optional<bool> {
-        currPath = path;
-        isSystemModule = isSystem;
-
-        switch (Kind) {
-        case SearchPathKind::Import:
-        case SearchPathKind::RuntimeLibrary: {
-          isFramework = false;
-
-          // On Apple platforms, we can assume that the runtime libraries use
-          // target-specifi module files wihtin a `.swiftmodule` directory.
-          // This was not always true on non-Apple platforms, and in order to
-          // ease the transition, check both layouts.
-          bool checkTargetSpecificModule = true;
-          if (Kind != SearchPathKind::RuntimeLibrary ||
-              !Ctx.LangOpts.Target.isOSDarwin()) {
-            auto modulePath = currPath;
-            llvm::sys::path::append(modulePath, genericModuleFileName);
-            llvm::ErrorOr<llvm::vfs::Status> statResult = fs.status(modulePath);
-
-            // Even if stat fails, we can't just return the error; the path
-            // we're looking for might not be "Foo.swiftmodule".
-            checkTargetSpecificModule = statResult && statResult->isDirectory();
-          }
+  SmallVector<std::string, 3> InterestingFilenames = {
+      (moduleName + ".framework").str(),
+      genericBaseName.getName(file_types::TY_SwiftModuleInterfaceFile),
+      genericBaseName.getName(file_types::TY_PrivateSwiftModuleInterfaceFile),
+      genericBaseName.getName(file_types::TY_SwiftModuleFile)};
+
+  auto searchPaths = Ctx.SearchPathOpts.moduleSearchPathsContainingFile(
+      InterestingFilenames, Ctx.SourceMgr.getFileSystem().get(),
+      Ctx.LangOpts.Target.isOSDarwin());
+  for (auto searchPath : searchPaths) {
+    currPath = searchPath->Path;
+    isSystemModule = searchPath->IsSystem;
+
+    switch (searchPath->Kind) {
+    case ModuleSearchPathKind::Import:
+    case ModuleSearchPathKind::RuntimeLibrary: {
+      isFramework = false;
+
+      // On Apple platforms, we can assume that the runtime libraries use
+      // target-specifi module files wihtin a `.swiftmodule` directory.
```

Comment:
I know you didn't actually change any of this but:
target-specifi -> target-specific
wihtin -> within